pkg/types/chain_reader.go

```diff
@@ -0,0 +1,64 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type chainReaderError string
+
+func (e chainReaderError) Error() string { return string(e) }
+
+const (
+	ErrInvalidType   = chainReaderError("invalid type")
+	ErrInvalidConfig = chainReaderError("invalid configuration")
+)
+
+func UnwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		if s.Code() == codes.Unimplemented {
+			return fmt.Errorf("%w : %s", errors.ErrUnsupported, s.String())
+		} else if s.Code() == codes.Unknown { // Only unwrap custom errors we return, leave alone any other gRPC generated error codes
+			return chainReaderError(s.String())
+		}
+	}
+	return err
+}
+
+type ChainReader interface {
+	// The params argument of GetLatestValue() can be any object which maps a set of generic parameters into chain specific parameters defined in RelayConfig. It must be a pointer which can be passed into
+	// json.Marshal(). Typically would be either an anonymous map such as `map[string]any{"baz": 42, "test": true}}` or something which implements the `MarshalJSON()` method (satisfying `Marshaller` interface).
```

Comment:
These are still missing info from the CLIP:
```suggestion
	// GetLatestValue gets the latest value....
	// The params argument can be any object which maps a set of generic parameters into chain specific parameters defined in RelayConfig. It must encode as an object via [json.Marshal].
	// Typically would be either an anonymous map such as `map[string]any{"baz": 42, "test": true}}`, a struct with `json` tags, or something which implements [json.Marshaler].
	//
	// Example use:
	//  type ProductParams struct {
	// 		Arg int `json:"arg"`
	//  }
	//  type ProductReturn struct {
	// 		Foo string `json:"foo"`
	// 		Bar *big.Int `json:"bar"`
	//  }
	//  func do(ctx context.Context, cr ChainReader) (resp ProductReturn, err error) {
	// 		err = cr.GetLatestValue(ctx, bc, "method", ProductParams{Arg:1}, &resp)
	// 		return
	//  }
```